File: Python/12-Tries(PrefixTrees)/Problems/2-WordSearch-II/WordSearch-II.py
from typing import List, Tuple
from collections import defaultdict

class TrieNode:
    """
    Represents a node in the Trie data structure.
    Each node stores a character and a dictionary of child nodes.
    """
    def __init__(self):
        self.children = defaultdict(TrieNode)
        self.is_word = False
        self.word = None  # Store the complete word at the leaf node

class Trie:
    """
    Represents a Trie (prefix tree) data structure.
    It is used to efficiently store and search for words.
    """
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the Trie.

        Args:
            word: The word to insert.
        """
        node = self.root
        for char in word:
            node = node.children[char]
        node.is_word = True
        node.word = word  # Store the word at the leaf

    def search(self, word: str) -> bool:
        """
        Searches for a word in the Trie.  (Full word search)

        Args:
            word: The word to search for.

        Returns:
            True if the word exists in the Trie, False otherwise.
        """
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_word

    def startsWith(self, prefix: str) -> bool:
        """
        Checks if there is any word in the trie that starts with the given prefix.

        Args:
            prefix: The prefix to search for.

        Returns:
            True if there is a word with the prefix, False otherwise.
        """
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True # Important: Return True, not node.is_word

class Solution:
    """
    Provides a solution to the "Word Search II" problem.
    Finds all words in a given board that are present in a word list.
    """

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        Finds all words in the board that are present in the word list.

        Args:
            board: A 2D list of characters representing the game board.
            words: A list of words to search for.

        Returns:
            A list of words found in the board.
        """
        if not board or not board[0] or not words:
            return []

        trie = Trie()
        for word in words:
            trie.insert(word)

        rows, cols = len(board), len(board[0])
        result = set()  # Use a set to avoid duplicate words

        def dfs(row: int, col: int, node: TrieNode, path: str) -> None:
            """
            Depth-first search (DFS) function to explore the board.

            Args:
                row: The current row index.
                col: The current column index.
                node: The current Trie node.
                path: The current path (string) being explored.  Not strictly needed,
                       but can be helpful for debugging or alternative solutions.
            """
            if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] not in node.children:
                return

            char = board[row][col]
            next_node = node.children[char]

            if next_node.is_word:
                result.add(next_node.word) # Add the word, not the path
                # Important: Do NOT return here.  There might be other words with this prefix.

            # Mark the current cell as visited to avoid revisiting it in the current path.
            temp = board[row][col]
            board[row][col] = '#'  # Use '#' or any character not in the board

            # Explore all four directions
            dfs(row + 1, col, next_node, path + char)
            dfs(row - 1, col, next_node, path + char)
            dfs(row, col + 1, next_node, path + char)
            dfs(row, col - 1, next_node, path + char)

            # Restore the original character of the cell after exploring its neighbors.
            board[row][col] = temp

        # Iterate through each cell in the board and start a DFS search.
        for r in range(rows):
            for c in range(cols):
                dfs(r, c, trie.root, "")

        return list(result)
    # Approach 1: Basic Trie + DFS (As implemented above)
    #    - Time Complexity: O(M * 4^L), where M is the number of cells in the board,
    #      and L is the maximum length of a word in the dictionary.
    #    - Space Complexity: O(N), where N is the total number of characters in all words.

    def findWords2(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        Slightly optimized:  Check for empty board *inside* the function.
        """
        if not board or not board[0] or not words:
            return []

        trie = Trie()
        for word in words:
            trie.insert(word)

        rows, cols = len(board), len(board[0])
        result = set()

        def dfs(row, col, node):
            if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] not in node.children:
                return
            char = board[row][col]
            next_node = node.children[char]

            if next_node.is_word:
                result.add(next_node.word)

            temp = board[row][col]
            board[row][col] = '#'
            dfs(row + 1, col, next_node)
            dfs(row - 1, col, next_node)
            dfs(row, col + 1, next_node)
            dfs(row, col - 1, next_node)
            board[row][col] = temp

        for r in range(rows):
            for c in range(cols):
                dfs(r, c, trie.root)
        return list(result)

    # Approach 2:  Slightly Cleaner DFS
    #   -  Same time and space complexity as Approach 1.
    #   -  Minor code cleanup in the DFS function.

    def findWords3(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
         Early termination in DFS if no matching word is possible
        """
        if not board or not board[0] or not words:
            return []

        trie = Trie()
        for word in words:
            trie.insert(word)

        rows, cols = len(board), len(board[0])
        result = set()

        def dfs(row, col, node):
            if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] not in node.children:
                return

            char = board[row][col]
            next_node = node.children[char]

            if next_node.is_word:
                result.add(next_node.word)

            temp = board[row][col]
            board[row][col] = '#'
            dfs(row + 1, col, next_node)
            dfs(row - 1, col, next_node)
            dfs(row, col + 1, next_node)
            dfs(row, col - 1, next_node)
            board[row][col] = temp

        for r in range(rows):
            for c in range(cols):
                if board[r][c] in trie.root.children: # Only start DFS if the first letter is in Trie
                    dfs(r, c, trie.root)
        return list(result)
    # Approach 3:  Early Termination in DFS
    #   - Time Complexity:  Potentially better than Approach 1 in some cases, but worst-case is still O(M * 4^L).
    #   - Space Complexity:  O(N)

    def findWords4(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        Optimization: Remove word from Trie after finding it.  This can significantly
        reduce the search space, especially if there are many overlapping words.
        """
        if not board or not board[0] or not words:
            return []

        trie = Trie()
        for word in words:
            trie.insert(word)

        rows, cols = len(board), len(board[0])
        result = set()

        def dfs(row, col, node):
            if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] not in node.children:
                return

            char = board[row][col]
            next_node = node.children[char]

            if next_node.is_word:
                result.add(next_node.word)
                # Remove word from Trie to avoid duplicate findings and reduce search space
                self.removeWord(trie.root, next_node.word) # Added removeWord
                return  # Important: return after finding a word

            temp = board[row][col]
            board[row][col] = '#'
            dfs(row + 1, col, next_node)
            dfs(row - 1, col, next_node)
            dfs(row, col + 1, next_node)
            dfs(row, col - 1, next_node)
            board[row][col] = temp

        for r in range(rows):
            for c in range(cols):
                if board[r][c] in trie.root.children:
                    dfs(r, c, trie.root)
        return list(result)

    def removeWord(self, root: TrieNode, word: str) -> None:
        """Removes a word from the Trie."""
        node = root
        path = []
        for char in word:
            if char not in node.children:
                return  # Word not in Trie
            path.append((node, char))
            node = node.children[char]

        if not node.is_word:
            return  # Word is not a complete word in Trie

        node.is_word = False
        node.word = None

        # Backtrack and remove nodes if they have no other children
        for i in range(len(path) - 1, -1, -1):
            parent, char = path[i]
            del parent.children[char]
            if parent.children:
                break # Stop if the parent has other children

    # Approach 4: Remove words from Trie after finding them
    #   - Time Complexity:  O(M * 4^L) in the worst case, but can be significantly faster in practice,
    #     especially when there are many overlapping words.  Removing words reduces the search space.
    #   - Space Complexity: O(N)  (Trie size can decrease, but is still bounded by O(N) initially)

    def findWords5(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        Optimized Trie Construction: Build Trie only with letters from the board.
        """
        if not board or not board[0] or not words:
            return []

        rows, cols = len(board), len(board[0])
        board_letters = set()
        for r in range(rows):
            for c in range(cols):
                board_letters.add(board[r][c])

        trie = Trie()
        for word in words:
            # Only insert words that can be formed from the board's letters
            if all(char in board_letters for char in word):
                trie.insert(word)

        result = set()

        def dfs(row, col, node):
            if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] not in node.children:
                return

            char = board[row][col]
            next_node = node.children[char]

            if next_node.is_word:
                result.add(next_node.word)
                self.removeWord(trie.root, next_node.word)
                return

            temp = board[row][col]
            board[row][col] = '#'
            dfs(row + 1, col, next_node)
            dfs(row - 1, col, next_node)
            dfs(row, col + 1, next_node)
            dfs(row, col - 1, next_node)
            board[row][col] = temp

        for r in range(rows):
            for c in range(cols):
                if board[r][c] in trie.root.children:
                    dfs(r, c, trie.root)
        return list(result)

    # Approach 5: Optimized Trie Construction
    #   - Time Complexity:  Potentially the most efficient, as it reduces Trie construction time
    #     and search space.  Still O(M * 4^L) in the worst case, but often much faster.
    #   - Space Complexity:  Can be significantly lower than O(N) if the board has few unique letters.
    #     In the worst case, it's still O(N).



if __name__ == "__main__":
    board = [
        ["o", "a", "a", "n"],
        ["e", "t", "a", "e"],
        ["i", "h", "k", "r"],
        ["i", "f", "l", "v"],
    ]
    words = ["oath", "pea", "eat", "rain"]

    solution = Solution()

    print("Approach 1: Basic Trie + DFS")
    result1 = solution.findWords(board, words)
    print(f"Found words: {result1}")  # Output: ['oath', 'eat']

    print("\nApproach 2: Slightly Cleaner DFS")
    result2 = solution.findWords2(board, words)
    print(f"Found words: {result2}")

    print("\nApproach 3: Early Termination in DFS")
    result3 = solution.findWords3(board, words)
    print(f"Found words: {result3}")

    print("\nApproach 4: Remove words from Trie after finding them")
    result4 = solution.findWords4(board, words)
    print(f"Found words: {result4}")

    print("\nApproach 5: Optimized Trie Construction")
    result5 = solution.findWords5(board, words)
    print(f"Found words: {result5}")

File: C++/12-Tries(PrefixTrees)/Problems/2-WordSearch-II/WordSearch-II.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>
#include <queue>
#include <map>

using namespace std;

// Approach 1: Brute Force (Time Limit Exceeded)
// For each word, iterate through the board and check if the word exists.
// Time Complexity: O(m * n * k * l), where m and n are board dimensions, k is the number of words, and l is the average word length.
// Space Complexity: O(1)
class Solution1 {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        int n = board[0].size();

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (dfs(board, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool dfs(vector<vector<char>>& board, int i, int j, string& word, int index) {
        if (index == word.length()) {
            return true;
        }

        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != word[index]) {
            return false;
        }

        char temp = board[i][j];
        board[i][j] = '#'; // Mark as visited to avoid revisiting in the same path.
        bool found =  dfs(board, i + 1, j, word, index + 1) ||
                       dfs(board, i - 1, j, word, index + 1) ||
                       dfs(board, i, j + 1, word, index + 1) ||
                       dfs(board, i, j - 1, word, index + 1);
        board[i][j] = temp; // Backtrack: Restore the original character.
        return found;
    }

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        vector<string> result;
        for (string word : words) {
            if (exist(board, word)) {
                result.push_back(word);
            }
        }
        return result;
    }
};

// Approach 2: Using HashSet (Optimization of Approach 1)
// Same as Brute Force but uses a HashSet to store the found words to avoid duplicates.
// Time Complexity: O(m * n * k * l), where m and n are board dimensions, k is the number of words, and l is the average word length.
// Space Complexity: O(min(k, m*n*l)),  where k is the number of words
class Solution2 {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        int n = board[0].size();

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (dfs(board, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool dfs(vector<vector<char>>& board, int i, int j, string& word, int index) {
        if (index == word.length()) {
            return true;
        }

        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != word[index]) {
            return false;
        }

        char temp = board[i][j];
        board[i][j] = '#';
        bool found =  dfs(board, i + 1, j, word, index + 1) ||
                       dfs(board, i - 1, j, word, index + 1) ||
                       dfs(board, i, j + 1, word, index + 1) ||
                       dfs(board, i, j - 1, word, index + 1);
        board[i][j] = temp;
        return found;
    }

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        unordered_set<string> result_set;
        for (string word : words) {
            if (exist(board, word)) {
                result_set.insert(word);
            }
        }
        return vector<string>(result_set.begin(), result_set.end());
    }
};

// Approach 3: Trie + DFS
// Build a Trie from the word list.  Then, perform DFS on the board, checking for Trie prefixes.
// Time Complexity: O(m * n * 4^L), where m and n are board dimensions, and L is the maximum word length.
// Space Complexity: O(N), where N is the total number of characters in all words.
struct TrieNode {
    vector<TrieNode*> children;
    string word; // Stores the complete word if this node is the end of a word.

    TrieNode() : children(26, nullptr), word("") {}

    ~TrieNode() {
        for (TrieNode* child : children) {
            if (child) {
                delete child;
            }
        }
    }
};

class Solution3 {
private:
    TrieNode* buildTrie(vector<string>& words) {
        TrieNode* root = new TrieNode();
        for (string& word : words) {
            TrieNode* node = root;
            for (char c : word) {
                int index = c - 'a';
                if (!node->children[index]) {
                    node->children[index] = new TrieNode();
                }
                node = node->children[index];
            }
            node->word = word; // Mark the end of the word.
        }
        return root;
    }

    void dfs(vector<vector<char>>& board, int i, int j, TrieNode* node, unordered_set<string>& result) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '#' || !node) {
            return;
        }

        char c = board[i][j];
        int index = c - 'a';
        TrieNode* nextNode = node->children[index];
        if (!nextNode) return;

        if (nextNode->word != "") { // Found a word
            result.insert(nextNode->word);
        }

        board[i][j] = '#'; // Mark as visited
        dfs(board, i + 1, j, nextNode, result);
        dfs(board, i - 1, j, nextNode, result);
        dfs(board, i, j + 1, nextNode, result);
        dfs(board, i, j - 1, nextNode, result);
        board[i][j] = c; // Backtrack
    }

public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode* root = buildTrie(words);
        unordered_set<string> result_set;
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                dfs(board, i, j, root, result_set);
            }
        }
        delete root;
        return vector<string>(result_set.begin(), result_set.end());
    }
};

// Approach 4: Trie + DFS (Optimized - No word storage in Trie)
// Optimized Trie approach.  Instead of storing the entire word in the TrieNode, check for word completion during DFS.
// Time Complexity: O(m * n * 4^L), where m and n are board dimensions, and L is the maximum word length.
// Space Complexity: O(N), where N is the total number of characters in all words.
struct TrieNode4 {
    vector<TrieNode4*> children;
    bool isWord;

    TrieNode4() : children(26, nullptr), isWord(false) {}
      ~TrieNode4() {
        for (TrieNode4* child : children) {
            if (child) {
                delete child;
            }
        }
    }
};

class Solution4 {
private:
    TrieNode4* buildTrie(vector<string>& words) {
        TrieNode4* root = new TrieNode4();
        for (string& word : words) {
            TrieNode4* node = root;
            for (char c : word) {
                int index = c - 'a';
                if (!node->children[index]) {
                    node->children[index] = new TrieNode4();
                }
                node = node->children[index];
            }
            node->isWord = true;
        }
        return root;
    }

    void dfs(vector<vector<char>>& board, int i, int j, TrieNode4* node, string& currentWord, unordered_set<string>& result, vector<string>& words) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '#' || !node) {
            return;
        }

        char c = board[i][j];
        int index = c - 'a';
        TrieNode4* nextNode = node->children[index];
        if (!nextNode) return;

        currentWord += c;
        if (nextNode->isWord) {
            result.insert(currentWord);
        }

        board[i][j] = '#';
        dfs(board, i + 1, j, nextNode, currentWord, result, words);
        dfs(board, i - 1, j, nextNode, currentWord, result, words);
        dfs(board, i, j + 1, nextNode, currentWord, result, words);
        dfs(board, i, j - 1, nextNode, currentWord, result, words);
        board[i][j] = c;
        currentWord.pop_back();
    }

public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode4* root = buildTrie(words);
        unordered_set<string> result_set;
        string currentWord = "";
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                dfs(board, i, j, root, currentWord, result_set, words);
            }
        }
        delete root;
        return vector<string>(result_set.begin(), result_set.end());
    }
};

// Approach 5: Trie + Optimized DFS (Early Stopping)
// Further optimization: Stop DFS traversal as soon as it's clear no word can be formed.
// Time Complexity: O(m * n * 4^L), where m and n are board dimensions, and L is the maximum word length.
// Space Complexity: O(N), where N is the total number of characters in all words.

struct TrieNode5 {
    vector<TrieNode5*> children;
    bool isWord;
    int count; // Keep track of the number of words with this prefix

    TrieNode5() : children(26, nullptr), isWord(false), count(0) {}
      ~TrieNode5() {
        for (TrieNode5* child : children) {
            if (child) {
                delete child;
            }
        }
    }
};

class Solution5 {
private:
    TrieNode5* buildTrie(vector<string>& words) {
        TrieNode5* root = new TrieNode5();
        for (string& word : words) {
            TrieNode5* node = root;
            for (char c : word) {
                int index = c - 'a';
                if (!node->children[index]) {
                    node->children[index] = new TrieNode5();
                }
                node = node->children[index];
                node->count++;
            }
            node->isWord = true;
        }
        return root;
    }

     void dfs(vector<vector<char>>& board, int i, int j, TrieNode5* node, string& currentWord, unordered_set<string>& result) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '#' || !node || node->count == 0) {
            return;
        }

        char c = board[i][j];
        int index = c - 'a';
        TrieNode5* nextNode = node->children[index];
        if (!nextNode) return;

        currentWord += c;
        if (nextNode->isWord) {
            result.insert(currentWord);
        }

        board[i][j] = '#';
        nextNode->count--; // Important: Decrement count as we traverse

        dfs(board, i + 1, j, nextNode, currentWord, result);
        dfs(board, i - 1, j, nextNode, currentWord, result);
        dfs(board, i, j + 1, nextNode, currentWord, result);
        dfs(board, i, j - 1, nextNode, currentWord, result);

        board[i][j] = c;
        currentWord.pop_back();
    }

public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode5* root = buildTrie(words);
        unordered_set<string> result_set;
        string currentWord = "";
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                dfs(board, i, j, root, currentWord, result_set);
            }
        }
        delete root;
        return vector<string>(result_set.begin(), result_set.end());
    }
};

int main() {
    vector<vector<char>> board = {
        {'o', 'a', 'a', 'n'},
        {'e', 't', 'a', 'e'},
        {'i', 'h', 'k', 'r'},
        {'i', 'f', 'l', 'v'}
    };
    vector<string> words = {"oath", "pea", "eat", "rain"};

    Solution1 sol1;
    vector<string> result1 = sol1.findWords(board, words);
    cout << "Approach 1: ";
    for (string word : result1) {
        cout << word << " ";
    }
    cout << endl;

    Solution2 sol2;
    vector<string> result2 = sol2.findWords(board, words);
    cout << "Approach 2: ";
    for (string word : result2) {
        cout << word << " ";
    }
    cout << endl;

    Solution3 sol3;
    vector<string> result3 = sol3.findWords(board, words);
    cout << "Approach 3: ";
    for (string word : result3) {
        cout << word << " ";
    }
    cout << endl;

    Solution4 sol4;
    vector<string> result4 = sol4.findWords(board, words);
    cout << "Approach 4: ";
    for (string word : result4) {
        cout << word << " ";
    }
    cout << endl;

    Solution5 sol5;
    vector<string> result5 = sol5.findWords(board, words);
    cout << "Approach 5: ";
    for (string word : result5) {
        cout << word << " ";
    }
    cout << endl;

    return 0;
}
